server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -341,7 +348,24 @@ public int indexShard(IndexRequest indexRequest) {
             checkNoRouting(indexRequest.routing());
             hash = hashSource(indexRequest);
             int shardId = hashToShardId(hash);
-            return (rerouteWritesIfResharding(shardId));
+            return rerouteWritesIfResharding(shardId);
+        }
+
+        @Override
+        public int rerouteToTarget(IndexRequest indexRequest) {
+            if (trackTimeSeriesRoutingHash) {
+                String routing = indexRequest.routing();
+                if (routing == null) {
+                    throw new IllegalStateException("Routing should be set by the coordinator");
+                }
+                return indexShard(indexRequest);
+            } else if (addIdWithRoutingHash) {
+                // TODO: is this correct?
```

Comment:
we're probably going to have to generate test cases for tsdb/logsdb

Reply:
https://elasticco.atlassian.net/browse/ES-13241